README.md
# Lab

[![Lab crate](https://img.shields.io/crates/v/lab.svg)](https://crates.io/crates/lab)
[![Lab documentation](https://docs.rs/lab/badge.svg)](https://docs.rs/lab)
![minimum rustc 1.31](https://img.shields.io/badge/rustc-1.36+-red.svg)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/TooManyBees/lab?branch=master&svg=true)](https://ci.appveyor.com/project/TooManyBees/lab)

Tools for converting RGB colors to L\*a\*b\* measurements.

RGB colors, for this crate at least, are considered to be composed of `u8`
values from 0 to 255, while L\*a\*b\* colors are represented by its own struct
that uses `f32` values.

# Usage

## Converting single values

To convert a single value, use one of the functions

* `lab::Lab::from_rgb(rgb: &[u8; 3]) -> Lab`
* `lab::Lab::from_rgba(rgba: &[u8; 4]) -> Lab` (drops the fourth alpha byte)
* `lab::Lab::to_rgb(&self) -> [u8; 3]`

```rust
extern crate lab;
use lab::Lab;

let pink_in_lab = Lab::from_rgb(&[253, 120, 138]);
// Lab { l: 66.639084, a: 52.251457, b: 14.860654 }
```

## Converting multiple values

To convert slices of values

* `lab::rgbs_to_labs(rgbs: &[[u8; 3]]) -> Vec<Lab>`
* `lab::labs_to_rgbs(labs: &[Lab]) -> Vec<[u8; 3]>`
* `lab::rgb_bytes_to_labs(bytes: &[u8]) -> Vec<Lab>`
* `lab::labs_to_rgb_bytes(labs: &[Lab]) -> Vec<u8>`

```rust
extern crate lab;
use lab::rgbs_to_labs;

let rgbs = vec![
    [0xFF, 0x69, 0xB6],
    [0xE7, 0x00, 0x00],
    [0xFF, 0x8C, 0x00],
    [0xFF, 0xEF, 0x00],
    [0x00, 0x81, 0x1F],
    [0x00, 0xC1, 0xC1],
    [0x00, 0x44, 0xFF],
    [0x76, 0x00, 0x89],
];

let labs = rgbs_to_labs(&rgbs);
```

```rust
extern crate lab;
use lab::rgb_bytes_to_labs;

let rgbs = vec![
    0xFF, 0x69, 0xB6,
    0xE7, 0x00, 0x00,
    0xFF, 0x8C, 0x00,
    0xFF, 0xEF, 0x00,
    0x00, 0x81, 0x1F,
    0x00, 0xC1, 0xC1,
    0x00, 0x44, 0xFF,
    0x76, 0x00, 0x89,
];

let labs = rgb_bytes_to_labs(&rgbs);
```

These functions will use x86_64 AVX2 instructions if compiled to a supported target.

## Minimum Rust version

Lab 0.8.0 requires Rust >= 1.36.0 for the [MaybeUninit](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html) struct

Lab 0.7.0 requires Rust >= 1.31.0 for the [chunks_exact](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks_exact) slice method


CHANGELOG.md
# 0.11.0

* bump `pretty_assertions` (dev dependency) crate from version 0.6 to 0.7
* bump `approx` (dev dependency) crate from version 0.3 to 0.5

# 0.10.0

* bump `rand` crate from version 0.5 to 0.8
* fix a potential edge case error. Per @mina86:
  > the `rgb_to_xyz_map` function uses threshold equal 10.  This is fine
  > if the argument is an integer in 0–255 range (255 * 0.0404 = 10.302)
  > which the function was originally designed to handle.  However, it
  > has since been changed to accept floating point numbers in said
  > range.  As a result, if a colour component is, say, 10.1, the
  > function will use the wrong gamma expansion formula.

# 0.9.0

* add `LCh` data struct - @mina86

# 0.8.2

* No code changes
* Includes Readme and license files in published crate

# 0.8.1

* Fix a bug in AVX2 path of `labs_to_rgb_bytes`

# 0.8.0

* Speed up AVX2 code path, primarily by bringing in an AVX2 impl of `powf`,
  `log`, and `exp`
* Remove `simd` as a public module
* `rgbs_to_labs` and `labs_to_rgbs`: uses AVX2 code paths when compiled on a
  target that supports it
* `rgb_bytes_to_labs` and `labs_to_rgb_bytes`: works with RGB triples in a flat
  `&[u8]` instead of `&[[u8; 3]]`

# 0.7.0

* add convenience methods for converting slices of labs/rgbs
* add experimental `simd` module with functions using avx and sse4.1 operations

# 0.6.0

* More accurate FP calculations - @mina86

# 0.5.0

* Update `rand` crate from version 0.3 to 0.5

# 0.4.4

Current version when `CHANGELOG.md` was created.


src/lib.rs
#![doc(html_root_url = "https://docs.rs/lab")]
#![deny(
    rust_2018_idioms,
    rust_2018_compatibility,
    future_incompatible,
    nonstandard_style,
    unused,
    missing_copy_implementations,
    missing_debug_implementations,
    clippy::all,
    clippy::cast_lossless,
    clippy::must_use_candidate,
    clippy::doc_markdown,
    clippy::unreadable_literal,
    clippy::unseparated_literal_suffix
)]
#![allow(clippy::excessive_precision)]

/*!

# Lab

Tools for converting RGB colors to L\*a\*b\* measurements.

RGB colors, for this crate at least, are considered to be composed of `u8`
values from 0 to 255, while L\*a\*b\* colors are represented by its own struct
that uses `f32` values.

# Usage

## Converting single values

To convert a single value, use one of the functions

* `lab::Lab::from_rgb(rgb: &[u8; 3]) -> Lab`
* `lab::Lab::from_rgba(rgba: &[u8; 4]) -> Lab` (drops the fourth alpha byte)
* `lab::Lab::to_rgb(&self) -> [u8; 3]`

```rust
extern crate lab;
use lab::Lab;

let pink_in_lab = Lab::from_rgb(&[253, 120, 138]);
// Lab { l: 66.639084, a: 52.251457, b: 14.860654 }
```

## Converting multiple values

To convert slices of values

* `lab::rgbs_to_labs(rgbs: &[[u8; 3]]) -> Vec<Lab>`
* `lab::labs_to_rgbs(labs: &[Lab]) -> Vec<[u8; 3]>`
* `lab::rgb_bytes_to_labs(bytes: &[u8]) -> Vec<Lab>`
* `lab::labs_to_rgb_bytes(labs: &[Lab]) -> Vec<u8>`

```rust
extern crate lab;
use lab::rgbs_to_labs;

let rgbs = vec![
    [0xFF, 0x69, 0xB6],
    [0xE7, 0x00, 0x00],
    [0xFF, 0x8C, 0x00],
    [0xFF, 0xEF, 0x00],
    [0x00, 0x81, 0x1F],
    [0x00, 0xC1, 0xC1],
    [0x00, 0x44, 0xFF],
    [0x76, 0x00, 0x89],
];

let labs = rgbs_to_labs(&rgbs);
```

```rust
extern crate lab;
use lab::rgb_bytes_to_labs;

let rgbs = vec![
    0xFF, 0x69, 0xB6,
    0xE7, 0x00, 0x00,
    0xFF, 0x8C, 0x00,
    0xFF, 0xEF, 0x00,
    0x00, 0x81, 0x1F,
    0x00, 0xC1, 0xC1,
    0x00, 0x44, 0xFF,
    0x76, 0x00, 0x89,
];

let labs = rgb_bytes_to_labs(&rgbs);
```

These functions will use x86_64 AVX2 instructions if compiled to a supported target.

## Minimum Rust version

Lab 0.7.0 requires Rust >= 1.31.0 for the [chunks_exact](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks_exact) slice method

*/

#[cfg(test)]
#[macro_use]
extern crate pretty_assertions;

#[cfg(test)]
mod approx_impl;

#[cfg(all(target_arch = "x86_64", target_feature = "avx2"))]
mod simd;

/// Struct representing a color in `CIALab`, a.k.a. L\*a\*b\*, color space
#[derive(Debug, PartialEq, Copy, Clone, Default)]
pub struct Lab {
    pub l: f32,
    pub a: f32,
    pub b: f32,
}

/// Struct representing a color in cylindrical `CIELCh` color space
#[derive(Debug, PartialEq, Copy, Clone, Default)]
pub struct LCh {
    pub l: f32,
    pub c: f32,
    pub h: f32,
}

// κ and ε parameters used in conversion between XYZ and La*b*.  See
// http://www.brucelindbloom.com/LContinuity.html for explanation as to why
// those are different values than those provided by CIE standard.
pub(crate) const KAPPA: f32 = 24389.0 / 27.0;
pub(crate) const EPSILON: f32 = 216.0 / 24389.0;
pub(crate) const CBRT_EPSILON: f32 = 6.0 / 29.0;

// S₀ and E₀ thresholds used in sRGB gamma.  The latter is scaled for encoded
// value in the range 0..255.
pub(crate) const S_0: f32 = 0.003_130_668_442_500_564;
pub(crate) const E_0_255: f32 = 3294.6 * S_0;

// Generated by srgb-matrices.py
const WHITE_X: f32 = 0.950_449_218_275_099_1;
const WHITE_Z: f32 = 1.088_916_648_430_471_5;

fn rgb_to_lab(r: u8, g: u8, b: u8) -> Lab {
    xyz_to_lab(rgb_to_xyz(r, g, b))
}

fn rgb_to_xyz(r: u8, g: u8, b: u8) -> [f32; 3] {
    rgb_to_xyz_inner(f32::from(r), f32::from(g), f32::from(b))
}

fn rgb_to_xyz_normalized(rgb: &[f32; 3]) -> [f32; 3] {
    rgb_to_xyz_inner(rgb[0] * 255.0, rgb[1] * 255.0, rgb[2] * 255.0)
}

#[inline(always)]
#[cfg(any(target_feature = "fma", test))]
fn mul3(a0: f32, a1: f32, a2: f32, b0: f32, b1: f32, b2: f32) -> f32 {
    a2.mul_add(b2, a1.mul_add(b1, a0 * b0))
}

#[inline(always)]
#[cfg(not(any(target_feature = "fma", test)))]
fn mul3(a0: f32, a1: f32, a2: f32, b0: f32, b1: f32, b2: f32) -> f32 {
    a0 * b0 + a1 * b1 + a2 * b2
}

#[inline]
fn rgb_to_xyz_inner(r: f32, g: f32, b: f32) -> [f32; 3] {
    #[inline]
    fn rgb_to_xyz_map(c: f32) -> f32 {
        if c > E_0_255 {
            const A: f32 = 0.055 * 255.0;
            const D: f32 = 1.055 * 255.0;
            ((c + A) / D).powf(2.4)
        } else {
            const D: f32 = 12.92 * 255.0;
            c / D
        }
    }

    let r = rgb_to_xyz_map(r);
    let g = rgb_to_xyz_map(g);
    let b = rgb_to_xyz_map(b);

    // Generated by srgb-matrices.py
    let x = mul3(
        r,
        g,
        b,
        0.412_410_846_488_538_8,
        0.357_584_567_852_951_9,
        0.180_453_803_933_608_33,
    );
    let y = mul3(
        r,
        g,
        b,
        0.212_649_342_720_652_83,
        0.715_169_135_705_903_8,
        0.072_181_521_573_443_33,
    );
    let z = mul3(
        r,
        g,
        b,
        0.019_331_758_429_150_258,
        0.119_194_855_950_983_97,
        0.950_390_034_050_337_3,
    );

    [x, y, z]
}

fn xyz_to_lab(xyz: [f32; 3]) -> Lab {
    #[inline]
    fn xyz_to_lab_map(c: f32) -> f32 {
        if c > EPSILON {
            c.powf(1.0 / 3.0)
        } else {
            (KAPPA * c + 16.0) / 116.0
        }
    }

    // It’s tempting to replace the division with a multiplication by inverse,
    // however that results in slightly worse test_grey_error benchmark.
    let x = xyz_to_lab_map(xyz[0] / WHITE_X);
    let y = xyz_to_lab_map(xyz[1]);
    let z = xyz_to_lab_map(xyz[2] / WHITE_Z);

    Lab {
        l: (116.0 * y) - 16.0,
        a: 500.0 * (x - y),
        b: 200.0 * (y - z),
    }
}

fn lab_to_xyz(lab: &Lab) -> [f32; 3] {
    let fy = (lab.l + 16.0) / 116.0;
    let fx = (lab.a / 500.0) + fy;
    let fz = fy - (lab.b / 200.0);
    let xr = if fx > CBRT_EPSILON {
        fx.powi(3)
    } else {
        ((fx * 116.0) - 16.0) / KAPPA
    };
    let yr = if lab.l > EPSILON * KAPPA {
        fy.powi(3)
    } else {
        lab.l / KAPPA
    };
    let zr = if fz > CBRT_EPSILON {
        fz.powi(3)
    } else {
        ((fz * 116.0) - 16.0) / KAPPA
    };

    [xr * WHITE_X, yr, zr * WHITE_Z]
}

fn xyz_to_rgb(xyz: [f32; 3]) -> [u8; 3] {
    let rgb = xyz_to_rgb_normalized(xyz);
    [
        (rgb[0] * 255.0).round() as u8,
        (rgb[1] * 255.0).round() as u8,
        (rgb[2] * 255.0).round() as u8,
    ]
}

fn xyz_to_rgb_normalized(xyz: [f32; 3]) -> [f32; 3] {
    let x = xyz[0];
    let y = xyz[1];
    let z = xyz[2];

    // Generated by srgb-matrices.py
    let r = mul3(
        x,
        y,
        z,
        3.240_812_398_895_283,
        -1.537_308_445_629_813_6,
        -0.498_586_522_906_966_6,
    );
    let g = mul3(
        x,
        y,
        z,
        -0.969_243_017_008_640_7,
        1.875_966_302_908_574_2,
        0.041_555_030_856_685_64,
    );
    let b = mul3(
        x,
        y,
        z,
        0.055_638_398_436_112_804,
        -0.204_007_460_932_413_62,
        1.057_129_570_286_143_4,
    );

    #[inline]
    fn xyz_to_rgb_map(c: f32) -> f32 {
        (if c > S_0 {
            1.055 * c.powf(1.0 / 2.4) - 0.055
        } else {
            12.92 * c
        })
        .min(1.0)
        .max(0.0)
    }

    [xyz_to_rgb_map(r), xyz_to_rgb_map(g), xyz_to_rgb_map(b)]
}

/// Convenience function to map a slice of RGB values to Lab values in serial
///
/// # Example
/// ```
/// # extern crate lab;
/// # use lab::{Lab, rgbs_to_labs};
/// let rgbs = &[[255u8, 0, 0], [255, 0, 255], [0, 255, 255]];
/// let labs = lab::rgbs_to_labs(rgbs);
/// assert_eq!(labs, vec![
///     Lab { l: 53.238235, a: 80.09231, b: 67.202095 },
///     Lab { l: 60.322693, a: 98.23698, b: -60.827957 },
///     Lab { l: 91.11428, a: -48.08274, b: -14.12958 }
/// ]);
/// ```
#[inline]
#[must_use]
pub fn rgbs_to_labs(rgbs: &[[u8; 3]]) -> Vec<Lab> {
    #[cfg(all(target_arch = "x86_64", target_feature = "avx2"))]
    let labs = simd::rgbs_to_labs(rgbs);

    #[cfg(not(all(target_arch = "x86_64", target_feature = "avx2")))]
    let labs = __scalar::rgbs_to_labs(rgbs);

    labs
}

/// RGB to Lab conversion that operates on a flat `&[u8]` of consecutive RGB triples.
///
/// # Example
/// ```
/// # extern crate lab;
/// # use lab::{Lab, rgb_bytes_to_labs};
/// let rgbs = &[255u8, 0, 0, 255, 0, 255, 0, 255, 255];
/// let labs = lab::rgb_bytes_to_labs(rgbs);
/// assert_eq!(labs, vec![
///     Lab { l: 53.238235, a: 80.09231, b: 67.202095 },
///     Lab { l: 60.322693, a: 98.23698, b: -60.827957 },
///     Lab { l: 91.11428, a: -48.08274, b: -14.12958 }
/// ]);
/// ```
#[must_use]
pub fn rgb_bytes_to_labs(bytes: &[u8]) -> Vec<Lab> {
    #[cfg(all(target_arch = "x86_64", target_feature = "avx2"))]
    let labs = simd::rgb_bytes_to_labs(bytes);

    #[cfg(not(all(target_arch = "x86_64", target_feature = "avx2")))]
    let labs = __scalar::rgb_bytes_to_labs(bytes);

    labs
}

/// Convenience function to map a slice of Lab values to RGB values in serial
///
/// # Example
/// ```
/// # extern crate lab;
/// # use lab::{Lab, labs_to_rgbs};
/// let labs = &[
///     Lab { l: 91.11321, a: -48.08751, b: -14.131201 },
///     Lab { l: 60.32421, a: 98.23433, b: -60.824894 },
///     Lab { l: 97.13926, a: -21.553724, b: 94.47797 },
/// ];
/// let rgbs = lab::labs_to_rgbs(labs);
/// assert_eq!(rgbs, vec![[0u8, 255, 255], [255, 0, 255], [255, 255, 0]]);
/// ```
#[inline]
#[must_use]
pub fn labs_to_rgbs(labs: &[Lab]) -> Vec<[u8; 3]> {
    #[cfg(all(target_arch = "x86_64", target_feature = "avx2"))]
    let rgbs = simd::labs_to_rgbs(labs);

    #[cfg(not(all(target_arch = "x86_64", target_feature = "avx2")))]
    let rgbs = __scalar::labs_to_rgbs(labs);

    rgbs
}

/// Lab to RGB conversion that returns RGB triples flattened into a `Vec<u8>`
///
/// # Example
/// ```
/// # extern crate lab;
/// # use lab::{Lab, labs_to_rgb_bytes};
/// let labs = &[
///     Lab { l: 91.11321, a: -48.08751, b: -14.131201 },
///     Lab { l: 60.32421, a: 98.23433, b: -60.824894 },
///     Lab { l: 97.13926, a: -21.553724, b: 94.47797 },
/// ];
/// let rgb_bytes = lab::labs_to_rgb_bytes(labs);
/// assert_eq!(rgb_bytes, vec![0, 255, 255, 255, 0, 255, 255, 255, 0]);
/// ```
#[inline]
#[must_use]
pub fn labs_to_rgb_bytes(labs: &[Lab]) -> Vec<u8> {
    #[cfg(all(target_arch = "x86_64", target_feature = "avx2"))]
    let bytes = simd::labs_to_rgb_bytes(labs);

    #[cfg(not(all(target_arch = "x86_64", target_feature = "avx2")))]
    let bytes = __scalar::labs_to_rgb_bytes(labs);

    bytes
}

#[doc(hidden)]
pub mod __scalar {
    use crate::rgb_to_lab;
    use crate::Lab;

    #[inline]
    #[must_use]
    pub fn labs_to_rgbs(labs: &[Lab]) -> Vec<[u8; 3]> {
        labs.iter().map(Lab::to_rgb).collect()
    }

    #[inline]
    #[must_use]
    pub fn labs_to_rgb_bytes(labs: &[Lab]) -> Vec<u8> {
        labs.iter()
            .map(Lab::to_rgb)
            .fold(Vec::with_capacity(labs.len() * 3), |mut acc, rgb| {
                acc.extend_from_slice(&rgb);
                acc
            })
    }

    #[inline]
    #[must_use]
    pub fn rgbs_to_labs(rgbs: &[[u8; 3]]) -> Vec<Lab> {
        rgbs.iter().map(Lab::from_rgb).collect()
    }

    #[inline]
    #[must_use]
    pub fn rgb_bytes_to_labs(bytes: &[u8]) -> Vec<Lab> {
        bytes
            .chunks_exact(3)
            .map(|rgb| rgb_to_lab(rgb[0], rgb[1], rgb[2]))
            .collect()
    }
}

impl Lab {
    /// Constructs a new `Lab` from a three-element array of `u8`s
    ///
    /// # Examples
    ///
    /// ```
    /// let lab = lab::Lab::from_rgb(&[240, 33, 95]);
    /// assert_eq!(lab::Lab { l: 52.334686, a: 75.55157, b: 19.995684 }, lab);
    /// ```
    #[must_use]
    pub fn from_rgb(rgb: &[u8; 3]) -> Self {
        rgb_to_lab(rgb[0], rgb[1], rgb[2])
    }

    #[doc(hidden)]
    #[must_use]
    pub fn from_rgb_normalized(rgb: &[f32; 3]) -> Self {
        xyz_to_lab(rgb_to_xyz_normalized(rgb))
    }

    /// Constructs a new `Lab` from a four-element array of `u8`s
    ///
    /// The `Lab` struct does not store alpha channel information, so the last
    /// `u8` representing alpha is discarded. This convenience method exists
    /// in order to easily measure colors already stored in an RGBA array.
    ///
    /// # Examples
    ///
    /// ```
    /// let lab = lab::Lab::from_rgba(&[240, 33, 95, 255]);
    /// assert_eq!(lab::Lab { l: 52.334686, a: 75.55157, b: 19.995684 }, lab);
    /// ```
    #[must_use]
    pub fn from_rgba(rgba: &[u8; 4]) -> Self {
        Lab::from_rgb(&[rgba[0], rgba[1], rgba[2]])
    }

    #[doc(hidden)]
    #[must_use]
    pub fn from_rgba_normalized(rgba: &[f32; 4]) -> Self {
        Lab::from_rgb_normalized(&[rgba[0], rgba[1], rgba[2]])
    }

    /// Returns the `Lab`'s color in RGB, in a 3-element array.
    ///
    /// # Examples
    ///
    /// ```
    /// let lab = lab::Lab { l: 52.330193, a: 75.56704, b: 19.989174 };
    /// let rgb = lab.to_rgb();
    /// assert_eq!([240, 33, 95], rgb);
    /// ```
    #[must_use]
    pub fn to_rgb(&self) -> [u8; 3] {
        xyz_to_rgb(lab_to_xyz(self))
    }

    #[doc(hidden)]
    #[must_use]
    pub fn to_rgb_normalized(&self) -> [f32; 3] {
        xyz_to_rgb_normalized(lab_to_xyz(self))
    }

    /// Measures the perceptual distance between the colors of one `Lab`
    /// and an `other`.
    ///
    /// # Examples
    ///
    /// ```
    /// # use lab::Lab;
    /// let pink = Lab { l: 66.6348, a: 52.260696, b: 14.850557 };
    /// let websafe_pink = Lab { l: 64.2116, a: 62.519463, b: 2.8871894 };
    /// assert_eq!(254.23636, pink.squared_distance(&websafe_pink));
    /// ```
    #[must_use]
    pub fn squared_distance(&self, other: &Lab) -> f32 {
        (self.l - other.l).powi(2) + (self.a - other.a).powi(2) + (self.b - other.b).powi(2)
    }
}

impl LCh {
    /// Constructs a new `LCh` from a three-element array of `u8`s
    ///
    /// # Examples
    ///
    /// ```
    /// let lch = lab::LCh::from_rgb(&[240, 33, 95]);
    /// assert_eq!(lab::LCh { l: 52.334686, c: 78.15284, h: 0.25873056 }, lch);
    /// ```
    #[must_use]
    pub fn from_rgb(rgb: &[u8; 3]) -> Self {
        LCh::from_lab(Lab::from_rgb(rgb))
    }

    /// Constructs a new `LCh` from a four-element array of `u8`s
    ///
    /// The `LCh` struct does not store alpha channel information, so the last
    /// `u8` representing alpha is discarded. This convenience method exists
    /// in order to easily measure colors already stored in an RGBA array.
    ///
    /// # Examples
    ///
    /// ```
    /// let lch = lab::LCh::from_rgba(&[240, 33, 95, 255]);
    /// assert_eq!(lab::LCh { l: 52.334686, c: 78.15284, h: 0.25873056 }, lch);
    /// ```
    #[must_use]
    pub fn from_rgba(rgba: &[u8; 4]) -> Self {
        LCh::from_lab(Lab::from_rgba(rgba))
    }

    /// Constructs a new `LCh` from a `Lab`
    ///
    /// # Examples
    ///
    /// ```
    /// let lab = lab::Lab { l: 52.33686, a: 75.5516, b: 19.998878 };
    /// let lch = lab::LCh::from_lab(lab);
    /// assert_eq!(lab::LCh { l: 52.33686, c: 78.15369, h: 0.25877 }, lch);
    ///
    /// let lab = lab::Lab { l: 52.33686, a: 0.0, b: 0.0 };
    /// let lch = lab::LCh::from_lab(lab);
    /// assert_eq!(lab::LCh { l: 52.33686, c: 0.0, h: 0.0 }, lch);
    /// ```
    #[must_use]
    pub fn from_lab(lab: Lab) -> Self {
        LCh {
            l: lab.l,
            c: lab.a.hypot(lab.b),
            h: lab.b.atan2(lab.a),
        }
    }

    /// Returns the `LCh`'s color in RGB, in a 3-element array
    ///
    /// # Examples
    ///
    /// ```
    /// let mut lch = lab::LCh { l: 52.33686, c: 78.15369, h: 0.25877 };
    /// assert_eq!([240, 33, 95], lch.to_rgb());
    ///
    /// lch.h += std::f32::consts::TAU;
    /// assert_eq!([240, 33, 95], lch.to_rgb());
    /// ```
    #[must_use]
    pub fn to_rgb(&self) -> [u8; 3] {
        self.to_lab().to_rgb()
    }

    /// Returns the `LCh`'s color in `Lab`
    ///
    /// Note that due to imprecision of floating point arithmetic, conversions
    /// between Lab and `LCh` are not stable.  A chain of Lab→LCh→Lab or
    /// LCh→Lab→LCh operations isn’t guaranteed to give back the source colour.
    ///
    /// # Examples
    ///
    /// ```
    /// let lch = lab::LCh { l: 52.33686, c: 78.15369, h: 0.25877 };
    /// let lab = lch.to_lab();
    /// assert_eq!(lab::Lab { l: 52.33686, a: 75.5516, b: 19.998878 }, lab);
    ///
    /// let lch = lab::LCh { l: 52.33686, c: 0.0, h: 0.25877 };
    /// let lab = lch.to_lab();
    /// assert_eq!(lab::Lab { l: 52.33686, a: 0.0, b: 0.0 }, lab);
    ///
    /// let inp = lab::Lab { l: 29.52658, a: 58.595745, b: -36.281406 };
    /// let lch = lab::LCh { l: 29.52658, c: 68.91881,  h: -0.5544043 };
    /// let out = lab::Lab { l: 29.52658, a: 58.59575,  b: -36.281406 };
    /// assert_eq!(lch, lab::LCh::from_lab(inp));
    /// assert_eq!(out, lch.to_lab());
    /// ```
    #[must_use]
    pub fn to_lab(&self) -> Lab {
        Lab {
            l: self.l,
            a: self.c * self.h.cos(),
            b: self.c * self.h.sin(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{labs_to_rgbs, rgbs_to_labs, LCh, Lab};
    use approx::assert_relative_eq;
    use rand::Rng;

    const PINK: Lab = Lab {
        l: 66.637_695,
        a: 52.250_145,
        b: 14.858_591,
    };

    #[rustfmt::skip]
    static COLOURS: [([u8; 3], Lab, LCh); 17] = [
        ([253, 120, 138],
         PINK,
         LCh { l: 66.637_695, c: 54.321_777, h: 0.277_060_2 }),

        ([127,   0,   0],
         Lab { l: 25.299_877, a: 47.77421, b: 37.752_514 },
         LCh { l: 25.299_877, c: 60.890_293, h: 0.668_753_86 }),
        ([  0, 127,   0],
         Lab { l: 45.87715, a: -51.405_922, b: 49.61748 },
         LCh { l: 45.87715, c: 71.445_526, h: 2.373_896 }),
        ([  0,   0, 127],
         Lab { l: 12.809_523, a: 47.237_186, b: -64.33636 },
         LCh { l: 12.809_523, c: 79.81553, h: -0.937_469_66 }),
        ([  0, 127, 127],
         Lab { l: 47.892_532, a: -28.680_845, b: -8.428_156 },
         LCh { l: 47.892_532, c: 29.893_557, h: -2.855_778_2 }),
        ([127,   0, 127],
         Lab { l: 29.525_677, a: 58.597_298, b: -36.28323 },
         LCh { l: 29.525_677, c: 68.92109, h: -0.554_415 }),
        ([255,   0,   0],
         Lab { l: 53.238_235, a: 80.09231, b: 67.202_095 },
         LCh { l: 53.238_235, c: 104.55094, h: 0.698_107_3 }),
        ([  0, 255,   0],
         Lab { l: 87.73554, a: -86.18078, b: 83.18251 },
         LCh { l: 87.73554, c: 119.776_695, h: 2.373_896 }),
        ([  0,   0, 255],
         Lab { l: 32.298_466, a: 79.192, b: -107.858_345 },
         LCh { l: 32.298_466, c: 133.8088, h: -0.937_469_66 }),
        ([  0, 255, 255],
         Lab { l: 91.11428, a: -48.08274, b: -14.12958 },
         LCh { l: 91.11428, c: 50.115_814, h: -2.855_778_7 }),
        ([255,   0, 255],
         Lab { l: 60.322_693, a: 98.23698, b: -60.827_957 },
         LCh { l: 60.322_693, c: 115.544_556, h: -0.554_414_87 }),
        ([255, 255,   0],
         Lab { l: 97.139, a: -21.556_675, b: 94.48001 },
         LCh { l: 97.139, c: 96.90801, h: 1.795_117_6 }),

        ([  0,   0,   0],
         Lab { l: 0.0, a: 0.0, b: 0.0 },
         LCh { l: 0.0, c: 0.0, h: 0.0 }),
        ([ 64,  64,  64],
         Lab { l: 27.09341, a: 0.0, b: 0.0 },
         LCh { l: 27.09341, c: 0.0, h: 0.0 }),
        ([127, 127, 127],
         Lab { l: 53.192_772, a: 0.0, b: 0.0 },
         LCh { l: 53.192_772, c: 0.0, h: 0.0 }),
        ([196, 196, 196],
         Lab { l: 79.15698, a: 0.0, b: 0.0 },
         LCh { l: 79.15698, c: 0.0, h: 0.0 }),
        ([255, 255, 255],
         Lab { l: 100.0, a: 0.0, b: 0.0 },
         LCh { l: 100.0, c: 0.0, h: 0.0 }),
    ];

    #[test]
    fn test_lab_from_rgb() {
        let expected: Vec<_> = COLOURS.iter().map(|(_, lab, _)| *lab).collect();
        let actual: Vec<_> = COLOURS
            .iter()
            .map(|(rgb, _, _)| Lab::from_rgb(rgb))
            .collect();
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_lab_to_rgb() {
        let expected: Vec<_> = COLOURS.iter().map(|(rgb, _, _)| *rgb).collect();
        let actual: Vec<_> = COLOURS.iter().map(|(_, lab, _)| lab.to_rgb()).collect();
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_lch_from_rgb() {
        let expected: Vec<_> = COLOURS.iter().map(|(_, _, lch)| *lch).collect();
        let actual: Vec<_> = COLOURS
            .iter()
            .map(|(rgb, _, _)| LCh::from_rgb(rgb))
            .collect();
        assert_relative_eq!(expected.as_slice(), actual.as_slice());
    }

    #[test]
    fn test_lch_to_rgb() {
        let expected: Vec<_> = COLOURS.iter().map(|(rgb, _, _)| *rgb).collect();
        let actual: Vec<_> = COLOURS.iter().map(|(_, _, lch)| lch.to_rgb()).collect();
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_lch_from_lab() {
        let expected: Vec<_> = COLOURS.iter().map(|(_, _, lch)| *lch).collect();
        let actual: Vec<_> = COLOURS
            .iter()
            .map(|(_, lab, _)| LCh::from_lab(*lab))
            .collect();
        assert_relative_eq!(expected.as_slice(), actual.as_slice());
    }

    #[test]
    fn test_lch_to_lab() {
        let mut expected: Vec<_> = COLOURS.iter().map(|(_, lab, _)| *lab).collect();
        let mut actual: Vec<_> = COLOURS.iter().map(|(_, _, lch)| lch.to_lab()).collect();

        // Floating point arithmetic is hard.  Due to accumulation of errors (or
        // perhaps imprecision of trig functions) the Lab→LCh→Lab conversion
        // produces slightly different colour than what the source.  Round a*
        // and b* to four decimal places to work around this.
        fn round(vec: &mut Vec<Lab>) {
            for lab in vec.iter_mut() {
                lab.a = (lab.a * 100_000.0).round() / 100_000.0;
                lab.b = (lab.b * 100_000.0).round() / 100_000.0;
            }
        }
        round(&mut expected);
        round(&mut actual);

        assert_eq!(expected, actual);
    }

    #[test]
    fn test_distance() {
        let ugly_websafe_pink = Lab {
            l: 64.2116,
            a: 62.519_463,
            b: 2.887_189_4,
        };
        assert!(254.65927 - PINK.squared_distance(&ugly_websafe_pink) <= f32::EPSILON);
    }

    #[test]
    fn test_send() {
        fn assert_send<T: Send>() {}
        assert_send::<Lab>();
    }

    #[test]
    fn test_sync() {
        fn assert_sync<T: Sync>() {}
        assert_sync::<Lab>();
    }

    #[test]
    fn test_rgb_to_lab_to_rgb() {
        let rgbs: Vec<[u8; 3]> = {
            let rand_seed = [1_u8; 32];
            let rng: rand::rngs::StdRng = rand::SeedableRng::from_seed(rand_seed);
            rng.sample_iter(&rand::distributions::Standard)
                .take(2048)
                .collect()
        };
        let labs = rgbs_to_labs(&rgbs);
        let rgbs2 = labs_to_rgbs(&labs);
        assert_eq!(rgbs2, rgbs);
    }

    #[test]
    fn test_grey_error() {
        // Grey colours have a* and b* components equal to zero.  This test goes
        // through all 8-bit greys and calculates squared error.  If it goes up,
        // a change might have worsen the precision of the calculations.  If it
        // goes down, calculations got better.
        let mut error: f64 = 0.0;
        let mut count: usize = 0;
        for i in 0..=255_u32 {
            let lab = Lab::from_rgb(&[i as u8, i as u8, i as u8]);
            if lab.a != 0.0 || lab.b != 0.0 {
                error = (f64::from(lab.a)).mul_add(f64::from(lab.a), error);
                error = (f64::from(lab.b)).mul_add(f64::from(lab.b), error);
                count += 1;
            }
        }
        assert_eq!((23, 10.627_054_791_711_998), (count, error * 1e9));
    }
}
